### PR TITLE
push notifications follow teammate name display user or server setting

### PR DESCRIPTION
**Pull request by enahum**
#### Summary
This PR addresses the teammate display name not being shown when a push notification was sent.

It should be tested and loadtested in order to ensure it does not creates performance issues, if that is the case we can always remove the user preference and always use the server one.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-11105

### CODE REVIEW

**Comment by jasonblais** [comment]
@enahum @esethna Can you help confirm the PR covers the following changes (we could get QA to test if needed):
1) Display name used when "First and last name" shown in Account Settings > Display > Teammate Name Display
2) Display name used for public/private channels
3) User preference always overwrites the system preference

@lieut-data FYI given mattermod fails to invoke a loadtest (ticket https://mattermost.atlassian.net/browse/MM-11126), will this block loadtesting changes in this PR, or is there a workaround until mattermod is fixed during this sprint?

**Comment by lieut-data** [comment]
@jasonblais, I'd be happy to test this one manually for now. Let me review the code first to see if we need loadtest changes first or not.

**Comment by lieut-data** [comment]
@jasonblais, it looks like this is in a code that's not currently loadtested, since we don't set `SendPushNotifications` and don't configure `PushNotificationServer` for loadtests. 

The core changes seem to result in just calling `GetPreferenceByCategoryAndNameForUser` more often than before. It's difficult to imagine how this would impact performance in a meaningful (or avoidable) way, so I'm not worried if we don't extend our loadtests right now to consider this path.

**Comment by jasonblais** [comment]
@lieut-data Thanks! Can you also confirm with Joram/Christopher since they had some concerns earlier.

Push notifications used to follow this setting and it was reverted due to performance concerns. But that was 12-18 months ago.

**Comment by lieut-data** [comment]
Good call, @jasonblais -- will do!

File: app/notification.go [hunk 1]
```diff
@@ -731,12 +720,25 @@ func (a *App) sendPushNotification(post *model.Post, user *model.User, channel *
...
+		channelName = fmt.Sprintf("@%v", senderName)
+	}
+
 	if contentsConfig != model.GENERIC_NO_CHANNEL_NOTIFICATION || channel.Type == model.CHANNEL_DIRECT {
 		msg.ChannelName = channelName
 	}
 
 	if ou, ok := post.Props["override_username"].(string); ok {
 		msg.OverrideUsername = ou
+		senderName = ou
```
Comment (hmhealey) [inline app/notification.go]: Do we need to check if we have username overrides enabled here, or is that done prior to this?

File: model/user.go [hunk 2]
```diff
@@ -406,11 +406,11 @@ func (u *User) AddNotifyProp(key string, value string) {
 }
 
 func (u *User) GetFullName() string {
-	if u.FirstName != "" && u.LastName != "" {
+	if len(u.FirstName) > 0 && len(u.LastName) > 0 {
```
Comment (hmhealey) [inline model/user.go]: Why change these to use len? As far as I can tell, this is equivalent to what we had before

Comment (enahum) [reply model/user.go]: true it is equivalent but by some strange reason it wasn't working, dunno the reason, but as soon as I did this change every option started working as expected

File: app/notification.go (hunk 1 above)
Comment (enahum) [reply app/notification.go]: I've added the check just to be sure, for the username and the icon

**Comment by lindalumitchell** [comment]
Thanks for the details, @esethna!

**Pull request merged**